core/includes/backup/backup.class.inc

```diff
@@ -0,0 +1,346 @@
+   *   by reference to allow it to be reassigned to a new file if needed.
+   *
+   * @return void
+   */
+  public function preRestore(BackupFile &$file) {
+    // No-op.
+  }
+
+  /**
+   * Change a file after restoration.
+   *
+   * This step is usually used to delete up the decompressed backup file.
+   *
+   * @param BackupFile $file
+   *  The file containing the backup that was restored. This is explicitly
+   *  passed by reference to allow it to be reassigned to a new file if needed.
+   *
+   * @return void
+   */
+  public function postRestore(BackupFile &$file) {
+    // No-op.
+  }
+
+  /**
+   * Get the form for the backup settings for this destination.
+   */
+  abstract public function backupSettingsForm();
+
+  /**
+   * Determine if this backup source is relevant based on the site configuration.
+   *
+   * @return boolean
+   */
+  //abstract public function applies();
+
+  /**
+   * Log a message that occurs during a backup or migration process.
+   *
+   * Often a log will not be able to be saved anywhere, since Backdrop may not
+   * be installed or in a working state, so log messages are held and can be
+   * retrieved and cleared as needed.
+   */
+  protected function log($message, $replacements, $log_type) {
+    $this->log[] = array($message, $replacements, $log_type);
+  }
+
+  /**
+   * Return the log for this backup instance.
+   *
+   * @return array
+   */
+  public function getLog() {
+    return $this->log;
+  }
+
+  /**
+   * Clear the log for this backup instance.
+   */
+  public function clearLog() {
+    $this->log = array();
+  }
+
+  /**
+   * Determines if time for a page execution has run out.
+   */
+  protected function timeoutCheck() {
+    // Max execution of 0 means unlimited.
+    if (ini_get('max_execution_time') == 0) {
+      return FALSE;
+    }
+    // Figure out when we should stop execution.
+    $backup_max_time = max(ini_get('max_execution_time'), $this->settings['backup_max_time']);
+    $timeout = (!empty($_SERVER['REQUEST_TIME']) ? $_SERVER['REQUEST_TIME'] : time()) + $backup_max_time - $this->settings['backup_timeout_buffer'];
+
+    return (time() > $timeout);
+  }
+
+  /**
+   * Compress a backup file with gzip.
+   *
+   * @return bool|BackupFile
+   *   The newly created compressed file as BackupFile object. FALSE if failure.
+   */
+  public function compress(BackupFile $file, $format = self::COMPRESSION_GZIP, $delete_original = FALSE) {
+    // Only gzip compression supported currently.
+    if ($format !== self::COMPRESSION_GZIP) {
+      return FALSE;
+    }
+
+    // If the file is already compressed, nothing to do here.
+    if ($file->lastExtension() === 'gz') {
+      return $file;
+    }
+
+    $success = FALSE;
+    $compressed_file = clone $file;
+    $compressed_file->pushExtension('gz');
+
+    $source = $file->filePath();
+    $destination = $compressed_file->filePath();
+    $level = 9;
+
+    if (@function_exists('gzopen')) {
```

Comment:
Why the `@` with function_exists()

AFAIK  function_exists() always returns a bool. I don't think, an error control operator is needed in this case.